### PR TITLE
[#23106] Add periodic.Sequence and periodic.Impulse transforms to Go SDK

### PR DESCRIPTION
**Pull request by hnnsgstfssn**
The new transforms extends support for the slowly updating side input pattern [1] as tracked in [2].

An attempt to mirror the logic of the Python implementation [3] has been made with minor idiomatic changes.

Java [4][5] and Python [6] have influenced the documentation and naming.

[1] https://beam.apache.org/documentation/patterns/side-inputs/
[2] https://github.com/apache/beam/issues/23106
[3] https://github.com/apache/beam/blob/v2.46.0/sdks/python/apache_beam/transforms/periodicsequence.py#L59
[4] https://beam.apache.org/releases/javadoc/2.46.0/org/apache/beam/sdk/transforms/PeriodicSequence.html
[5] https://beam.apache.org/releases/javadoc/2.46.0/org/apache/beam/sdk/transforms/PeriodicImpulse.html
[6] https://beam.apache.org/releases/pydoc/2.46.0/apache_beam.transforms.periodicsequence.html?highlight=periodicimpulse#module-apache_beam.transforms.periodicsequence

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by Abacn** [comment]
Thanks for the contribution! This is a very useful transform. For the drain issue, Python and Java implementation also did not drain until recent fix (#23765 #25716). Basically the transform should (i) update watermark when advance and (ii) truncate immediately when truncate is called (happens in drain)

**Comment by lostluck** [comment]
Thanks for your patience! I'm still getting caught up from being on vacation. I'm intending to get this review done today or tomorrow.

**Comment by lostluck** [comment]
This is failing our Staticheck action. Not hard to fix at least.

https://github.com/apache/beam/actions/runs/4392151772/jobs/7693061461?pr=25808

```
Error: transforms/periodic/periodic.go:61:18: calling math.Ceil on a converted integer is pointless (SA4015)
Error: transforms/periodic/periodic.go:132:6: func withNowFunc is unused (U1000)
Error: Process completed with exit code 1.
```

File: sdks/go/pkg/beam/transforms/periodic/periodic.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+	return &sdf.ManualWatermarkEstimator{}
+}
+
+func (fn *sequenceGenDoFn) ProcessElement(ctx context.Context, we *sdf.ManualWatermarkEstimator, rt *sdf.LockRTracker, sd SequenceDefinition, emit func(beam.EventTime, int64)) (sdf.ProcessContinuation, error) {
+	currentOutputIndex := rt.GetRestriction().(offsetrange.Restriction).Start
+	currentOutputTimestamp := sd.Start.Add(sd.Interval * time.Duration(currentOutputIndex))
+	currentTime := fn.now()
+	we.UpdateWatermark(currentOutputTimestamp)
+	for currentOutputTimestamp.Before(currentTime) {
+		if rt.TryClaim(currentOutputIndex) {
```
Comment (lostluck) [inline sdks/go/pkg/beam/transforms/periodic/periodic.go:95]: Just to explain what happens here and how it works:

The transform will continue until the restriction is fully processed, or the runner has told the bundle to split. On that split, tryClaim returns false, GetError nil, isDone false, and then we reschedule for later with the ResumeProcessingIn continuation. Note that this isn't a hard timer, and a runner may be early or late (usually early).

The output watermark is held to the last output timestamp. When Stop happens, the watermark can then advance to whereever the source is at (probably end of Global Window).

File: sdks/go/pkg/beam/transforms/periodic/periodic.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+func Impulse(s beam.Scope, start, end time.Time, interval time.Duration, opts ...ImpulseOption) beam.PCollection {
+	if interval <= 0 {
+		interval = math.MaxInt64
+	}
+	if start.IsZero() {
+		start = time.Now()
+	}
+	if start.After(end) {
+		start = end
+	}
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/transforms/periodic/periodic.go:162]: And mTime.Times are prevented from being serialized. 

I'd use int64s makes it clear, and it's an implementation detail. (and did so for a similar, but test specific transform, for testing prism). Reduces likely hood of architecture specific issues too since int's width is arch dependent.

TBH it's odd to me that the "value" for the periodic impulses are time instants, which are the event time anyway. I'd prefer we avoid the needless duplication Java and Python take by just having int for Sequence, and []byte{} for Impulse. Int is specific and clear enough for sequence (it's a sequence of numbers, done), but int64 is likely better for clarity on range across architectures. []byte is what the normal Impulse returns, and so it would make it more compatible / drop in with existing transforms that consume the []byte, even if it's ignored.

We can also just encode the time/timestamp into the []byte, if we so choose, but again, it's redundant.

File: sdks/go/pkg/beam/transforms/periodic/periodic.go
Comment (lostluck) [inline sdks/go/pkg/beam/transforms/periodic/periodic.go:1]: We should be able to add a short unit test if using the `prism` runner directly. 

While the runner isn't fully complete yet, it *does* run and execute ProcessContinuation transforms and watermarks!

It just doesn't do the splitting just yet, or actually "wait" for any process continuations at the moment. But when the "sequence" is done, it will terminate, so we can add a test with period of a second, a duration of a minute, and then count that we're getting 60 elements out of the transform. (Small risk of getting 59 instead, as a flake...)

File: sdks/go/pkg/beam/transforms/periodic/periodic.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/transforms/periodic/periodic.go:162]: For handling End, we can also just have a variable set to "largest practical time", and use that. It's going to be a very rare case where someone wants to process a stream of data that extends past the year 9999.  (Which we could work around by changing the serialization... but that's not urgent now.) 

We can also pick something like another arbitrary time with a nanosecond scale difference as a sentinel value that we check for (eg. Zero + 1 nanosecond), which we check for and treat as End of Time. Beam doesn't promise handling event time granularities under microseconds IIRC (if smaller than milliseconds at all...)

**Review by lostluck — CHANGES_REQUESTED** [review]
This is great! Thanks! I do have comments. Thanks for such a valuable first contribution to Beam! :) 

I have a few commons on example formatting and noting that we should use the new Prism runner to be able to have unit test pipelines for periodic. I think I've covered your other questions too.

File: sdks/go/examples/slowly_updating_side_input/slowly_updating_side_input.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+			periodic.Impulse(
+				s,
+				startTime,
+				endTime,
+				periodicSequenceInterval,
+			),
+		),
+		beam.Trigger(trigger.Repeat(trigger.Always())),
+		beam.PanesDiscard(),
+	)
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/examples/slowly_updating_side_input/slowly_updating_side_input.go]: WRT the cache:

Beam's execution abstraction, the FnAPI, has provisions for caching data cross bundle from the StateAPI on the SDK side, in order to avoid repeated deserialization, and additional round trips to the primary store of the runner to fetch state data. Side inputs also come across the StateAPI. In particular, very valuable for streaming jobs, as typically a single "SDK harness" is usually responsible for the same key all the time, so for tight windows it would look up the same data from the side inputs.

---

WRT not applying after the impulse change. Makes sense since an int64 was being received from the "update" function.

File: sdks/go/pkg/beam/transforms/periodic/periodic.go
Comment (lostluck) [reply sdks/go/pkg/beam/transforms/periodic/periodic.go:1]: The direct runner (and basically all non-portable runners), are not great, since they mis-align expectations for users when they move to executing on a "real" runner, like Flink or Dataflow. There's also no "spec" about what a Direct Runner should do to properly test things, so the Go Direct Runner is missing a number of features that the Java and Python ones use.

Essentially Prism is going to replace the Go Direct Runner as the default runner for the Go SDK at least, and hopefully make testing and local runs for all facets of Beam easier. The prism Readme has a bunch of the vision, and desired goals, but I gave a talk at last year's Beam Summit, about the motivations, especially in all the "hidden" bits of beam that users usually don't interact with, but are affected by.

https://www.youtube.com/watch?v=G4lbkvAG6xk

File: sdks/go/pkg/beam/transforms/periodic/periodic.go [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+	return []offsetrange.Restriction{rest}
+}
+
+// TruncateRestriction immediately truncates the entire restrication.
+func (fn *sequenceGenDoFn) TruncateRestriction(_ *sdf.LockRTracker, _ SequenceDefinition) offsetrange.Restriction {
+	return offsetrange.Restriction{}
+}
+
+func (fn *sequenceGenDoFn) CreateWatermarkEstimator() *sdf.ManualWatermarkEstimator {
+	return &sdf.ManualWatermarkEstimator{}
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/transforms/periodic/periodic.go:79]: While I agree in principle, if we were to take that route we should add a constructor function to the sdf package that does that instead `NewManualWatermarkEstimator()`, so it's clearer where to start, and avoid others needing to think about it as well.

On the the other hand, the Unix Epoch was 50+ years ago, and users that would be trying to reprocess data dated older than that would know they're probably going to need some special handling around it, and as you've said, it can be worked around already. I'd say, we do nothing at this juncture.

**Comment by lostluck** [comment]
The drain panic is outside of your code. From [here]( https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/core/runtime/exec/sdf_invokers.go#L320) which means it's invoking a Drain for a KV for some reason...  But sequenceGenDoFn isn't a KV, it does have restrictions though.

OK the thing that's failing here is that a recent change to support context.Context as a parameter into all the SDF methods changed how this one executes.

In particular, it moved from determining what to pass into the function based on its signature to based on what's input. But that's not how this Transform needs to work.

"old code"
https://github.com/apache/beam/blame/9fb9d2cb415e903e4222021f8848bf4420157b2a/sdks/go/pkg/beam/core/runtime/exec/sdf_invokers.go#L337

Ooof. This is why I want prism to implement all of Beam... including Drains, so this stuff can be tested without involving Dataflow.

--------

The fix is to change the trInvoker method a little, since it's generally passing in the wrong thing anyway. https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/core/runtime/exec/sdf.go#L296

The problem we're running into is deciding whether the Main Input is a KV or not, so the output can be expanded out. This is something unique to the Go SDK since we didn't have generics to lean on in the original design to have a "real" KV.

Basically, we are seeing a *FullValue{ Elm: Input, Elm2: Restriction}.  

So the right place to change is actually what's passed into the invoker in exec/sdf.go. 

There's already handling code there... Let me think about this for a bit, since I need to follow what's happening in the code...

File: sdks/go/pkg/beam/transforms/periodic/periodic_test.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	p, s := beam.NewPipelineWithRoot()
+	sd := SequenceDefinition{
+		Interval: time.Second,
+		Start:    0,
+		End:      time.Minute.Milliseconds(),
+	}
+	in := beam.Create(s, sd)
+	out := Sequence(s, in)
+	passert.Count(s, out, "SecondsInMinute", 60)
+	beam.Init()
```
Comment (lostluck) [inline sdks/go/pkg/beam/transforms/periodic/periodic_test.go:38]: Two things:
First, `beam.Init()` should be called before pipeline construction, since it's what enables the binary to behave as either a worker or a main. 

Second, calling beam.Init in tests isn't recommended as a rule since it can/will get called multiple times. 

So we have the PTest package to call it after test binary flag parsing, by overriding the default TestMain.

If you add the following:
```
func TestMain(m *testing.M) {
	os.Exit(ptest.MainRetWithDefault(m, "prism"))
}
```

You can even then trigger pipeline execution with `ptest.RunAndValidate(t, p)`, which will fail if an error is returned, but also check any counters from the returned pipeline result, which is another way we could validate this behavior other than `passert.Count`.

**Review by lostluck — CHANGES_REQUESTED** [review]

File: sdks/go/pkg/beam/transforms/periodic/periodic.go [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+	return rest.Size()
+}
+
+func (fn *sequenceGenDoFn) SplitRestriction(_ SequenceDefinition, rest offsetrange.Restriction) []offsetrange.Restriction {
+	return []offsetrange.Restriction{rest}
+}
+
+// TruncateRestriction immediately truncates the entire restrication.
+func (fn *sequenceGenDoFn) TruncateRestriction(_ *sdf.LockRTracker, _ SequenceDefinition) offsetrange.Restriction {
+	return offsetrange.Restriction{}
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/transforms/periodic/periodic.go:75]: Looking at the "default" implementation which does the same, a `nil` is probably the best actually, since it will abort processing immeadiately.  

The default implementation is to handle unbounded restrictions, but is sort of a little less obvious due to the Estimator involved.

[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/transforms/periodic/periodic.go:75]: Oh, but we can't do nil here. I think returning a zero value restriction is correct. 

We don't want to return the exact same restruction since we want this to terminate, and if the restriction isn't shortened somehow, it will never drain.

File: sdks/go/pkg/beam/transforms/periodic/periodic.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/transforms/periodic/periodic.go:125]: (3) That's also very good. SGTM.

**Comment by lostluck** [comment]
It's very pleasing that we can get coverage for this! https://app.codecov.io/gh/apache/beam/pull/25808

Can't help on the drain side yet, but we'll get there.

**Comment by lostluck** [comment]
OK, I think if this is patched in, it should resolve the drain issue. (Specifically the sdf.go file, the test doesn't need changing).

https://github.com/apache/beam/pull/25908

If that works, please review/approve/lgtm that chagne and I'll merge it in, and we can unblock Periodic impulse!

File: sdks/go/examples/slowly_updating_side_input/slowly_updating_side_input.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/examples/slowly_updating_side_input/slowly_updating_side_input.go]: I didn't address the larger side input window. Since it's an example, we don't need to over configure things. I'd make the window 5 times larger to demonstrate the paradigm, and an explicit comment that's what the larger window is for.

**Comment by lostluck** [comment]
Other example is merged in. Let me know when the test changes are made and I'll do one last pass. If it's good then,  I'll merge this in!

Thanks again for your contribution and patience.

File: sdks/go/pkg/beam/transforms/periodic/periodic.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+	register.Emitter1[[]byte]()
+	beam.RegisterType(reflect.TypeOf(SequenceDefinition{}))
+}
+
+// SequenceDefinition holds the configuration for generating a sequence of
+// timestamped elements at an interval.
+type SequenceDefinition struct {
+	Interval time.Duration
+	Start    int64
+	End      int64
```
Comment (lostluck) [inline sdks/go/pkg/beam/transforms/periodic/periodic.go:47]: Please add comments that these represent times in milliseconds.

Also would be good to add a construction function to do this conversion for users from `time.Time` if they're using `periodic.Sequence` themselves rather than through `periodic.Impulse`.

**Review by lostluck — APPROVED** [review]
Just one suggestion around documentation of SequenceDefinition and a helper constructor function for it and otherwise LGTM.

Thanks again!

**Review by lostluck — APPROVED** [review]
Thank you very much! This is very exciting.
